### DISCUSSION TITLE
Extending a dialect to include additional create table clauses

### DISCUSSION THREAD

**Discussion by mitches-got-glitches** [sqlalchemy/sqlalchemy#12188] [Usage Questions] [answered]
I need some guidance on usage around being able to use dialect specific clauses when creating table - I'm struggling to get what I need from the documentation, Github issues and Stack Overflow, so I would really appreciate the help!

My issue is specific to the Databricks dialect so I will use that as my example, but I'm hoping the solution is more generalised. When doing a `CREATE TABLE` on Databricks, there are [optional table clauses that you can define](https://docs.databricks.com/en/sql/language-manual/sql-ref-syntax-ddl-create-table-using.html#create-table-[using]) to change creation behaviour. The one that I'm interested in particularly is to specify a `LOCATION` with a `path` variable.

My question is: how can I make use of dialect specific table clauses in SQL Alchemy?

I've gone into the Databricks dialect and I can see that they have [a `post_create_table`  method](https://github.com/databricks/databricks-sql-python/blob/680b3b6df55518bb22939a9c1f052759c1b680df/src/databricks/sqlalchemy/_ddl.py#L17-L27) in their `DatabricksDDLCompiler` class which inherits from SQL Alchemy's `compiler.DDLCompiler`. [This method is called within the `visit_create_table` method](https://github.com/sqlalchemy/sqlalchemy/blob/main/lib/sqlalchemy/sql/compiler.py#L6666) of the inherited `DDLCompiler` and [`post_create_table` is left blank](https://github.com/sqlalchemy/sqlalchemy/blob/da0a3e1c3ec200cf489697f179559fa490bbd01a/lib/sqlalchemy/sql/compiler.py#L6922-L6923) in `sqlalchemy`, this I assume is because it is a default for specific dialects to override (what the Databricks dialect does) - code here:

```python
class DatabricksDDLCompiler(compiler.DDLCompiler):
    def post_create_table(self, table):
        post = [" USING DELTA"]
        if table.comment:
            comment = self.sql_compiler.render_literal_value(
                table.comment, sqltypes.String()
            )
            post.append("COMMENT " + comment)

        post.append("TBLPROPERTIES('delta.feature.allowColumnDefaults' = 'enabled')")
        return "\n".join(post)
```

My approach so far has been to create my own class which inherits from `DatabricksDDLCompiler` and override `post_create_table` to add the additional table clauses that I want - following their pattern for `comment`, the one difference being that `comment` is already an attribute of a `sqlalchemy.Table` object whereas my attribute is not:

```python
from databricks.sqlalchemy._ddl import DatabricksDDLCompiler
from databricks.sqlalchemy.base import DatabricksDialect

from sqlalchemy.sql import sqltypes


class MyDatabricksDDLCompiler(DatabricksDDLCompiler):
    def post_create_table(self, table):
        post = [" USING DELTA"]
        if table.comment:
            comment = self.sql_compiler.render_literal_value(
                table.comment, sqltypes.String()
            )
            post.append("COMMENT " + comment)
        # `_databricks_location` doesn't exist as a table attribute so it fails at this line.
        if table._databricks_location:
            location = self.sql_compiler.render_literal_value(
                table.location, sqltypes.String()
            )
            post.append("LOCATION " + location)

        post.append("TBLPROPERTIES('delta.feature.allowColumnDefaults' = 'enabled')")
        return "\n".join(post)

# Override the DDL compiler on the dialect with my own
DatabricksDialect.ddl_compiler = MyDatabricksDDLCompiler
```

The override of the dialect works, but my issue is that I'm not entirely sure how to pass additional arguments through to this `post_create_table` method. Trying to access from the `table` argument doesn't work.

When I create my model (I'm using `SQLModel`) - I'm trying to feed `_databricks_location` through to `__table_args__`, but I'm guessing that because `_databricks_location` is not an accepted argument of the `Table` constructor it's just being lost. I've seen that dialect specific arguments can be specified here (for example with Postgres), so what am I missing here? I feel like I'm close but I can't quite figure out the final piece of the puzzle. Is there an easier way than what I'm doing?

```python
from sqlmodel import SQLModel

class MyModel(SQLModel, table=True):
    # Defines the schema - we will create this in the next step.
    __table_args__ = {
        "schema": SCHEMA,
        "extend_existing": True,
        "_databricks_location": "abfss://container@https_path/my_model",
    }
```

**Comment by zzzeek** [comment, marked as answer]
> My question is: how can I make use of dialect specific table clauses in SQL Alchemy?

So this is something that would be better as part of the databricks dialect directly so when you do your code you'll want to submit a PR to them for a permanent change to their dialect.   short of that if you are looking for something that works locally you'll want to use the [compiles extension](https://docs.sqlalchemy.org/en/20/core/compiler.html).

first you'll want to add your argument to the dialect's [construct_arguments](https://github.com/sqlalchemy/sqlalchemy/blob/da0a3e1c3ec200cf489697f179559fa490bbd01a/lib/sqlalchemy/engine/interfaces.py#L1055) collection, databricks does not seem to have one so you can make a new one   (here's an [example](https://github.com/sqlalchemy/sqlalchemy/blob/da0a3e1c3ec200cf489697f179559fa490bbd01a/lib/sqlalchemy/dialects/mysql/base.py#L2552) of a populated one)

```py

from databricks.sqlalchemy.base import DatabricksDialect
DataBricksDialect.construct_arguments = [
       (sa_schema.Table, {"location": None}),
]

```

from there you can add `databricks_location='xyz'` to your `Table` definition.    you can extract it from the `Table` object like `my_table.dialect_options['databricks']['location']`.


then as far as the ddl_compiler thing you are doing, you can do that, but it's simpler to use `@compiles` like this:

```py

from sqlalchemy.ext.compiler import compiles
from sqlalchemy.sql.schema import CreateTable

@compiles(CreateTable, "databricks")
def _create_table(element, compiler, **kw):
     create_table_text = compiler.visit_create_table(element, **kw)
     create_table_text += "YOUR TEXT HERE"   # or use re.sub() to insert text in a specific place
     return create_table_text
```

**Reply by mitches-got-glitches** [reply]
Thanks for your prompt reply Mike (@zzzeek) - I appreciate it! I've been scanning and rescanning the Custom SQL compilations page because I knew that that was the path the docs were directing me to for this problem, and I've had some success with it in the past, but I couldn't quite see the way forward, so thanks for the extra context.

Yeah, I definitely agree about getting this in the databricks dialect so the full featured functionality is out-of-the-box. I'll open an issue on their GitHub page linking to this discussion to get their feedback.

I'll give your implementation a try for my own needs now, and hopefully the dialect will be expanded in due time.

**Reply by mitches-got-glitches** [reply]
I have got it working! 🙌 For completion and to just make some final clarifications:

```python
# This first import wasn't totally clear from the example given.
import sqlalchemy.schema as sa_schema
from databricks.sqlalchemy.base import DatabricksDialect

DatabricksDialect.construct_arguments = [
    (sa_schema.Table, {"location": None}),
]
```

```python
from sqlalchemy.ext.compiler import compiles
# I have imported CreateTable from this module as it didn't seem to exist in the
# `sqlalchemy.sql.schema` module given. Works with this.
from sqlalchemy.sql.ddl import CreateTable


@compiles(CreateTable, "databricks")
def _create_table(element, compiler, **kw):
    create_table_text = compiler.visit_create_table(element, **kw)
    # The table object seems to be accessible from the `.element` attribute.
    location = element.element.dialect_options["databricks"].get("location")
    if location:
        lines = create_table_text.split("\n")
        new_clause = f"LOCATION '{location}'"
        # create_table_text ends with two newlines, I found it easiest to split by "\n"
        # insert where I wanted into the list and then join back on "\n" 
        create_table_text = "\n".join(lines[:-2] + [new_clause] + lines[-2:])
    return create_table_text
```
```python
class MyModel(SQLModel, table=True):
    __table_args__ = {
        # Argument passed through without the preceding underscore.
        "databricks_location": "abfss://container@https_path/my_model",
    }
```
